File: zenscript-code-model/src/main/java/raylras/zen/model/type/ClassType.java
package raylras.zen.model.type;

import raylras.zen.model.symbol.ClassSymbol;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolProvider;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Deque;

public record ClassType(ClassSymbol symbol) implements Type, SymbolProvider {

    public String getQualifiedName() {
        return symbol.getQualifiedName();
    }

    public String getSimpleName() {
        return symbol.getSimpleName();
    }

    @Override
    public String getTypeName() {
        return getQualifiedName();
    }

    @Override
    public String getSimpleTypeName() {
        return getSimpleName();
    }

    @Override
    public boolean isSuperclassTo(Type type) {
        if (type instanceof ClassType that) {
            Deque<ClassSymbol> deque = new ArrayDeque<>();
            deque.push(that.symbol);
            while (!deque.isEmpty()) {
                ClassSymbol pop = deque.pop();
                if (pop.equals(this.symbol)) {
                    return true;
                }
                deque.addAll(pop.getInterfaces());
            }
        }
        return false;
    }

    @Override
    public Collection<Symbol> getSymbols() {
        MemberValidator validator = new MemberValidator();
        validator.addAll(symbol.getDeclaredMembers());
        Deque<ClassSymbol> interfaceDeque = new ArrayDeque<>(symbol.getInterfaces());
        while (!interfaceDeque.isEmpty()) {
            ClassSymbol pop = interfaceDeque.pop();
            validator.addAll(pop.getSymbols());
            interfaceDeque.addAll(pop.getInterfaces());
        }
        return validator.getMembers();
    }

    @Override
    public String toString() {
        return getTypeName();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/ClassType.kt
package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.ClassSymbol
import raylras.zen.model.symbol.FunctionSymbol
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.SymbolProvider
import java.util.*

data class ClassType(val symbol: ClassSymbol) : Type, SymbolProvider {
    override val typeName: String by symbol::qualifiedName
    override val simpleTypeName: String by symbol::simpleName

    override fun isSupertypeTo(type: Type): Boolean {
        if (type is ClassType) {
            val deque: Deque<ClassSymbol> = ArrayDeque()
            deque.push(type.symbol)
            while (deque.isNotEmpty()) {
                val pop = deque.pop()
                if (pop == this.symbol) {
                    return true
                }
                deque.addAll(pop.interfaces)
            }
        }
        return false
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        val validator = MemberValidator()
        symbol.declaredMembers.forEach { validator.add(it) }
        val interfaceDeque: Deque<ClassSymbol> = ArrayDeque(symbol.interfaces.toList())
        while (interfaceDeque.isNotEmpty()) {
            val pop = interfaceDeque.pop()
            pop.getSymbols().forEach { validator.add(it) }
            interfaceDeque.addAll(pop.interfaces)
        }
        return validator.getMembers().asSequence()
    }

    override fun toString() = typeName
}

fun ClassType?.firstAnonymousFunctionOrNull(): FunctionSymbol? {
    return this?.symbol?.declaredMembers
        ?.filterIsInstance<FunctionSymbol>()
        ?.firstOrNull { it.simpleName.isEmpty() }
}
